415 configurable settings

roschaefer opened the pull request:
**🚨 Before we deploy this feature to production, we should make sure that the onboarding for the #Schweinesystem project is completed as Isa already sent out invites using the old system.**

close #415 

merge this after #445

roschaefer commented:
@tillprochaska could you give me a review on this?

![download (18)](https://user-images.githubusercontent.com/2110676/96630280-0db2e180-1315-11eb-8344-366a2e4d041e.png)

tillprochaska commented on app/views/onboarding/index.html.erb:
```diff
@@ -1,36 +1,18 @@
 <% end %>
 
 <main>
-  <%= c 'onboarding_hero',
-    photo: '/onboarding/hero.jpg',
-    heading: I18n.t('onboarding.heading')
-  %>
```
I don’t really understand why you removed the page title from this component, thus relying on negative margins?

I guess that’s because you wanted to store all the page content’s in a single setting? I’d make an exception for the page title here for multiple reasons:

* I don’t really like applying negative, random margins to elements. In general, I try not to apply margins to components directly, as margins are external styles which in many cases can have unwanted side effects which makes the component context-dependant. Although I think, in this specific case, this might be dismissible, as it’s unlikely that we’re going to reuse the onboarding in other contexts.

* More importantly, I think it makes sense to store the page title separately, as we probably should use it as the HTML document title as well. (This isn’t something we’re doing at the moment, but we should do in the future. See #443)

tillprochaska commented on config/environments/test.rb:
```diff
@@ -14,7 +14,7 @@
 ENV['TELEGRAM_BOT_API_KEY'] = 'TOKEN'
 ENV['TELEGRAM_BOT_USERNAME'] = 'TestingBot'
-ENV['HUNDRED_EYES_PROJECT_NAME'] = '50survivors'
+ENV['HUNDRED_EYES_PROJECT_NAME'] = 'TestingProject'
```
I think `100eyes` (or even `Unbenanntes Projekt`) would be a better default than `TestingProject`, especially for the future, when real clients will actually use the software.

roschaefer replied:
It's `test` environment, clients won't see that.

tillprochaska commented on config/locales/de.yml:
```diff
@@ -27,7 +26,7 @@ de:
         other: Recherchen
     hero:
-      title: Willkommen bei 100eyes. Nutze die Intelligenz der Community.
+      title: Willkommen bei %{project_name}. Nutze die Intelligenz der Community.
```
The same applies here as above – the software editors are using is still "100eyes", only the project they are using 100eyes for has a name.

tillprochaska commented on config/locales/de.yml:
```diff
@@ -101,8 +100,49 @@ de:
+        label: Absender
+      inbound_email_password:
+        label: Inbound Passwort
+        help: Bei Sendgrid "%{callback_url}" unter "Inbound Parse" eintragen
```
I think this could be a more helpful, friendlier help text, fort example:

> Trage diese URL in den Sendgrid-Einstellungen unter „Inbound Parse” ein. Ein detaillierte Anleitung findest du in der [Sendgrid-Dokumentation](https://sendgrid.com/docs/for-developers/parsing-email/setting-up-the-inbound-parse-webhook/).

tillprochaska commented on config/locales/de.yml:
```diff
@@ -101,8 +100,49 @@ de:
+        label: Inbound Passwort
+        help: Bei Sendgrid "%{callback_url}" unter "Inbound Parse" eintragen
+      onboarding_page:
+        label: Willkommens-Seite
```
I think just "Willkommens-Seite" might be a bit ambiguous. I think in other places of the UI we will be using the Term "Einladungs-Link", so I’d go for "Einladungs-Seite" or even just "Onboarding-Seite".

tillprochaska commented on config/locales/de.yml:
```diff
@@ -229,25 +269,6 @@ de:
       Herzlich Willkommen bei %{project_name}. Danke, dass Sie an unserer Dialog-Recherche teilnehmen. In den kommenden Wochen nutzen wir Telegram für den Austausch. Ein Bot arbeitet wie ein Postbote für uns und stellt Fragen und Antworten zu. Ihre Nachrichten lesen und beantworten wir aber selbstverständlich persönlich. Wir freuen uns sehr auf den Dialog mit Ihnen!
```
This would need to be configurable as well as the welcome message and especially the names of the team members will likely change from project to project.

tillprochaska commented on config/routes.rb:
```diff
@@ -35,4 +35,8 @@
     end
   end
+
+  namespace :admin do
```
Is there any reason for nesting this inside of the `admin` namespace? Why not simply `/settings`.

roschaefer replied:
I sticked to the documentation of https://github.com/huacnlee/rails-settings-cached here. Similar to `telegram/webhook_controller.rb` which could be `telegram_controller.rb` as well.

tillprochaska reviewed (commented):
I haven’t had the time to check this out and have a look at the UI so far, but from what I could see reading through the code, looks good in general to me.

I think an important feature would be adding file uploads for the project logo and hero image, although we could keep it as it is right now (and host those resources externally), if that’d would cost too much time.

tillprochaska commented on app/models/setting.rb:
```diff
@@ -0,0 +1,38 @@
+  # field :confirmable_enable, default: "0", type: :boolean
+  # field :admin_emails, default: "admin@rubyonrails.org", type: :array
+  # field :omniauth_google_client_id, default: (ENV["OMNIAUTH_GOOGLE_CLIENT_ID"] || ""), type: :string, readonly: true
+  # field :omniauth_google_client_secret, default: (ENV["OMNIAUTH_GOOGLE_CLIENT_SECRET"] || ""), type: :string, readonly: true
```
Just saw you already removed that in the meantime! 🎈

tillprochaska commented on docker-compose.override.yml:
```diff
@@ -16,17 +18,13 @@ x-defaults: &x-defaults
-    <<: *x-defaults
-    command: ./bin/webpack-dev-server
-    ports: [ '3035:3035' ]
-
```
Just ignore that, I wrote that comment yesterday evening.

tillprochaska commented on app/components/hero/hero.html.erb:
```diff
@@ -7,7 +7,7 @@
         <%= c 'stack' do %>
           <%= c 'heading', style: :hero do %>
-            <%= I18n.t 'components.hero.title' %>
+            <%= I18n.t 'components.hero.title', project_name: Setting.project_name %>
```
See also these related comments:

https://github.com/tactilenews/100eyes/pull/434#discussion_r509458750
https://github.com/tactilenews/100eyes/pull/434#discussion_r508766015

drjakob replied:
file uploads for the project logo and hero image would be a great idea

tillprochaska replied:
> file uploads for the project logo and hero image would be a great idea

I added the feature request as a separate ticket over here: #450

roschaefer commented on app/controllers/admin/settings_controller.rb:
```diff
@@ -0,0 +1,26 @@
+  class SettingsController < ApplicationController
+    def index; end
+
+    def create
```
That's because RESTful route for PATCH requires an `:id`. There is no such thing as a *single* resource in rails: https://guides.rubyonrails.org/routing.html#resource-routing-the-rails-default

The documentation of `rails-cached-settings` is using `create` for that reason: https://github.com/huacnlee/rails-settings-cached#how-to-manage-settings-in-the-admin-interface

tillprochaska replied:
Okay, makes sense. You can still just define routes explicitly, you don’t have to follow those convention if you feel they aren’t a good fit for the use case.

```
get '/settings', to: 'settings#index'
post '/settings', to: 'settings#update'

roschaefer replied:
..and then also Rails form helpers don't work out-of-the-box :disappointed: I tried exactly that before :confused:

tillprochaska replied:
I just tried this out locally and it worked without any problems (at least I was able to update settings and didn’t notice any difference.)

I used the route definitions I mentioned above and changed the form action to `settings_path` (instead of `admin_settings_path`).

tillprochaska replied:
I just tried this out locally and it worked without any problems (at least I was able to update settings and didn’t notice any difference.)

I used the route definitions I mentioned above and changed the form action to `settings_path` (instead of `admin_settings_path`).

tillprochaska commented on app/controllers/admin/settings_controller.rb:
```diff
@@ -0,0 +1,31 @@
+
+    private
+
+    def setting_params
```
Why do you use the singular `setting` here? My understanding ist that a setting is an individual configuration option, e.g. `project_name`.

roschaefer replied:
No special reason except that most people name those things after the singular of the activerecord model it's supposed to change. So, `user_params`, `post_params`, not `users_params`. Feel free to change that! That shouldn't have any side-effects.

tillprochaska reviewed (changes requested):
Just checked out the UI for the settings. I think I’d go for a plain vertical layout for the form fields so we can add help texts to clarify what each of the settings is for.

<img width="1140" alt="Screen Shot 2020-10-23 at 12 43 43" src="https://user-images.githubusercontent.com/1512805/96995134-00247400-152e-11eb-9f22-c2b22b2ec493.png">

If you agree, I can make those changes…

roschaefer commented:
@tillprochaska agree with what you said. I already committed a couple of suggestions. Feel free to pluralize `SettingForm` and `setting_params` if you want.

tillprochaska commented:
@roschaefer I changed a few UI related things and also added the UI texts for the onboarding success page to the settings. (Will fix those rubocop errors later.) It’s quite long already, at some point we should probably think about splitting the sections up into separate pages (similar to the way settings are organised e.g. in the GitHub UI).

This is what the page now looks like. Maybe either @FrauCsu or @drjakob could have a look at it and check if everything that needs to be configurable by editors is in there?

![localhost_3000_admin_settings(iPad Pro) (1)](https://user-images.githubusercontent.com/1512805/97020278-a5047880-1551-11eb-9b8a-8cb07755b56f.png)

roschaefer commented:
> @roschaefer I changed a few UI related things and also added the UI texts for the onboarding success page to the settings. (Will fix those rubocop errors later.) It’s quite long already, at some point we should probably think about splitting the sections up into separate pages (similar to the way settings are organised e.g. in the GitHub UI).
> 
> This is what the page now looks like. Maybe either @FrauCsu or @drjakob could have a look at it and check if everything that needs to be configurable by editors is in there?
> 
> ![localhost_3000_admin_settings(iPad Pro) (1)](https://user-images.githubusercontent.com/1512805/97020278-a5047880-1551-11eb-9b8a-8cb07755b56f.png)

I don't see all text fields yet, that's because you didn't push them, right?

drjakob commented on app/components/nav_bar/nav_bar.html.erb:
```diff
@@ -3,7 +3,7 @@
       <li class="NavBar-item">
         <%= link_to root_path, class: 'NavBar-logo' do %>
-          <%= Rails.configuration.project_name %>
+          <%= Setting.project_name %>
```
Probably we could find a combination: e.g. 100eyes:Schweinesystem

tillprochaska reviewed (approved):
Hey @roschaefer, I implemented the remaining changes I suggested in my previous review. Apart from that, I introduced a `Markdown` component (similar to the `Plaintext` component we already have) and [renamed the controller action/routes](https://github.com/tactilenews/100eyes/pull/434#discussion_r508769403) we discussed here. (The change didn’t break any tests and I also verified it sill works manually.)

Maybe have a look at those changes to make sure I didn’t do any nonsense.

As I already commented above, please make sure not to merge this PR yet. We need to merge #429 before that!
